remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java

```diff
@@ -0,0 +1,172 @@
+            addRemoteLogSegmentMetadata(updatedMetadata);
+        }
+
+        if (targetState == RemoteLogSegmentState.DELETE_SEGMENT_FINISHED) {
+            log.debug("Cleaning up the state for : [{}]", metadataUpdate);
+            // remove this entry when the state is moved to delete_segment_finished
+            Map<Integer, Long> leaderEpochs = existingMetadata.segmentLeaderEpochs();
+            for (Map.Entry<Integer, Long> entry : leaderEpochs.entrySet()) {
+                NavigableMap<Long, RemoteLogSegmentId> offsetToIds = leaderEpochToOffsetToId.get(entry.getKey());
+                // remove the mappings where this segment is deleted.
+                offsetToIds.values().remove(remoteLogSegmentId);
+            }
+
+            // remove the segment-id mapping.
+            idToSegmentMetadata.remove(remoteLogSegmentId);
+        }
+    }
+
+    public Iterator<RemoteLogSegmentMetadata> listAllRemoteLogSegments() {
+        ArrayList<RemoteLogSegmentMetadata> list = new ArrayList<>(idToSegmentMetadata.values());
+        list.addAll(remoteLogSegmentIdInProgress.stream().map(id -> idToSegmentMetadata.get(id))
+                .collect(Collectors.toList()));
+        list.sort(Comparator.comparingLong(RemoteLogSegmentMetadata::startOffset));
+        return list.iterator();
+    }
+
+    public Iterator<RemoteLogSegmentMetadata> listRemoteLogSegments(int leaderEpoch) {
+        NavigableMap<Long, RemoteLogSegmentId> map = leaderEpochToOffsetToId.get(leaderEpoch);
+        return map != null ? map.values().stream().map(id -> idToSegmentMetadata.get(id)).iterator()
+                           : Collections.emptyIterator();
+    }
+
+    public Optional<Long> highestLogOffset(int leaderEpoch) {
+        NavigableMap<Long, RemoteLogSegmentId> offsetToSegmentId = leaderEpochToOffsetToId.get(leaderEpoch);
+        if (offsetToSegmentId == null) {
+            return Optional.empty();
+        }
+
+        long max = 0L;
+        for (RemoteLogSegmentId id : offsetToSegmentId.values()) {
+            RemoteLogSegmentMetadata metadata = idToSegmentMetadata.get(id);
+            Map.Entry<Integer, Long> nextEntry = metadata.segmentLeaderEpochs().higherEntry(leaderEpoch);
+            // If there is an higher entry than the given leader epoch, end-offset of the given leader epoch is,
+            // (next leader epoch's start-offset) -1.
+            long nextVal = nextEntry != null ? nextEntry.getValue() - 1 : metadata.endOffset();
+            max = Math.max(max, nextVal);
+        }
+
+        return Optional.of(max);
+    }
+
+    /**
+     * This will be added to copy_in_progress metadata list. This will be removed from that list once it is moved to the
+     * next state which can be COPY_SEGMENT_FINISHED or DELETE_SEGMENT_STARTED.
+     *
+     * @param remoteLogSegmentMetadata
+     */
+    public void addToInProgress(RemoteLogSegmentMetadata remoteLogSegmentMetadata) {
```

Comment:
Before we insert into the map/set, we should check if the provided `remoteLogSegmentMetadata.state()` is `COPY_SEGMENT_STARTED`.